lib/repositories/survey_repository.dart

```diff
@@ -0,0 +1,47 @@
+import 'package:survey/models/survey_info.dart';
+import 'package:survey/services/api/survey/survey_api_service.dart';
+import 'package:survey/services/local_storage/local_storage_service.dart';
+import 'package:survey/services/locator/locator_service.dart';
+
+abstract class SurveyRepository {
+  static const listLocalStorageKey = "survey_repository_list";
+
+  Future<bool> get isSurveysCached;
+
+  Future<List<SurveyInfo>> fetchSurveys({bool force});
+}
+
+class SurveyRepositoryImpl implements SurveyRepository {
+  final SurveyApiService _surveyApiService = locator.get();
+  final LocalStorageService _localStorageService = locator.get();
+
+  @override
+  Future<bool> get isSurveysCached async {
+    final items = await _localStorageService
+        .getListObject<SurveyInfo>(SurveyRepository.listLocalStorageKey);
+    return items != null;
+  }
+
+  @override
+  Future<List<SurveyInfo>> fetchSurveys({bool force = false}) async {
+    // Refer items from local storage
+    if (!force) {
+      final items = await _localStorageService
+          .getListObject<SurveyInfo>(SurveyRepository.listLocalStorageKey);
+      if (items != null) {
+        return items;
+      }
+    }
+
+    // Fetch from server
+    final list = await _surveyApiService.list(params: SurveyListParams());
+
+    // Storage it
+    _localStorageService.setListObject(
+      list.items,
+      key: SurveyRepository.listLocalStorageKey,
+    );
+
+    return list.items;
```

Comment:
with this approach, you're actually just returning either: Local (cached) data OR Remote API data (fetch). The requirement from the IC said that we should serve the local data first, and then if there is new content from the fetching API, we can update the content according to that. Since we are in the scope of this PR which doesn't require such implementation, I'd propose 2 ways for your future usage/ref:

#1: change the return type of `fetchSurveys` into an `Observable` stream, so you can emit 2 times: 1 is local data, and 2 is API data, by this way, you can `listen` just 1 time and if there is data updates, it can display correctly from the UI.
#2: keep this as-is, but change the flag name a little bit: `force` -> `isCached`; so you return 2 different types of data, and let the lower layer when using this function decide the logic to fetch from both local + API.

Reply:
@sleepylee Hmm, I don't see that requirement in [here](https://docs.google.com/document/u/1/d/e/2PACX-1vQERKjvXwKo3GpwRY4Qjc7U0afxjnPRcYEj6VK-scOEiZOYkolirtopUR8cZ2U65akiujAd0LKGp0ZS/pub)
Only them
```
- The list of surveys must be fetched when opening the application and cached onto the device. Show a loading animation when fetching the list of surveys.
- The cache of the list of surveys can be refreshed from the home screen by doing a pull-to-refresh. Show a loading animation when refreshing the list of surveys.
```

Btw, this requirement makes me confused
> The list of surveys must be fetched when opening the application and cached onto the device. Show a loading animation when fetching the list of surveys.

If we always fetch when opening the application, so what is the purpose of cached there?

Reply:
it is called: `Offline-first approach` in case you haven't heard about this design, @markgravity 

https://hasura.io/blog/design-guide-to-offline-first-apps/
https://developer.chrome.com/docs/apps/offline_apps/

Basically, we cached the data, and on the next time we enter the screen again, we basically serve the local data whenever possible, this reduces the amount of network call when unnecessary; also providing a good UX behavior because the data displays immediately rather than fetching from the API all the time.

Reply:
@markgravity I'll update the requirement in the IC a little bit so it will be clear for everyone later, sorry for the confusion 🙇

Reply:
@sleepylee got u. Let me update it!
Btw, this requirement should be updated, cause it kindly generic and vague 😁

Reply:
I added this section, hope it is clearer now on the requirement.
![image](https://user-images.githubusercontent.com/13435717/120747058-968c2f80-c52a-11eb-91ce-af2a42a4d6f4.png)